### PR TITLE
File upload using android-async-http

### PR DESCRIPTION
**Pull request by tobiaselindberg**
Added another http-client called android-async-http to handle file uploads to our API. Also added user agent to both volley and android-async-http since the sdk previously did not get our user agent injected
to the sdk.

### CODE REVIEW

File: sdk/src/main/java/com/podio/sdk/provider/FileProvider.java [hunk 1]
```diff
@@ -21,16 +21,16 @@
...
  *
  * @author László Urszuly
  */
 public class FileProvider extends Provider {
 
-    static class Path extends Filter {
+    public static class FileFilter extends Filter {
 
-        protected Path() {
-            super("file/v2");
```
Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/provider/FileProvider.java]: why this change from `file/v1` to `file/` ?

File: sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    protected AndroidAsyncHttpRequest(AsyncHttpClient client, Context context, String url, File file, Class<T> resultType) {
+        this.client = client;
+        this.context = context;
+        this.url = url;
+        this.file = file;
+        this.classOfResult = resultType;
+        this.callbackManager = new CallbackManager();
+        if (isSessionAboutToExpire()) {
```
Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: 2 things:
* if the session is about to expire, we refresh the token... and do not trying performing a request?
* performing a request in the constructor is kind of awkward anyway. Constructors are there to initialize objects, I would force the user to call `performRequest` himself, as this seems very anti intuitive.

File: sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        if (error != null) {
+            throw error;
+        }
+
+        return result;
+    }
+
+    @Override
+    public Request<T> withResultListener(ResultListener<T> contentListener) {
+        callbackManager.addResultListener(contentListener, isDone, result);
```
Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: in the error listener you do `isDone && error != null`
this is the listener for any result? or a successful result? if the later, should check the `error` variable as well?

File: sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
+/*
```
Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: I know that this is all over the place, 
but I would remove the Copyright note. This is 90's practice. Now you put in the `README.md` if you want and that's it.

File: sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpClient.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Override
+    @Deprecated
+    public Request<Void> forceRefreshTokens() {
+        throw new UnsupportedOperationException("AndroidAsyncHttpClient only supports uploading to our Files API.");
+    }
+
+    @Override
+    public <T> Request<T> request(Request.Method method, Filter filter, Object item, Class<T> classOfResult) {
+        if (!(filter instanceof FileProvider.FileFilter) || method != Request.Method.POST || !(item instanceof File)) {
```
Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpClient.java]: a bit fussy, but the conditional can be simplified taking out all negations and reversing the logic:
`if ((filter instanceof FileProvider.FileFilter) && method == Request.Method.POST && (item instanceof File)) {
    return request
} else {
   throw exception
}`

That's somewhat easier to read

File: sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java (hunk 4 above)
Comment (RabieJradi) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: I would be happy for that change. One less thing to remember when writing code.

File: sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+                @Override
+                public void onFailure(int statusCode, Header[] headers, Throwable throwable, String rawJsonData, T errorResponse) {
+                    if (throwable instanceof ConnectTimeoutException || throwable instanceof SocketTimeoutException) {
+                        error = new NoResponseError(throwable);
+                    } else if (rawJsonData != null) {
+                        try {
+                            error = new ApiError(rawJsonData, statusCode, throwable);
+                        } catch (JsonSyntaxException jsonSyntaxException) {
+                            error = new PodioError(throwable);
```
Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: I assume that this is how SDK works, but doesn't feel like we should have `ApiError` *and* `PodioError`.
What is the difference? Just the payload?
Not something to fix now, but just a note...

File: sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpClient.java (hunk 5 above)
Comment (tobiaselindberg) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpClient.java]: Will change that.

File: sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java (hunk 4 above)
Comment (tobiaselindberg) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: I will leave this as is for now since that is how it is in all files.

File: sdk/src/main/java/com/podio/sdk/Podio.java [hunk 7]
```diff
@@ -227,10 +230,11 @@ public static SessionListener removeGlobalSessionListener(SessionListener sessio
      * @param clientSecret
      *         The corresponding Podio client secret.
      *
-     * @see Podio#setup(Context, String, String, String, String, SSLSocketFactory)
+     * @see Podio#setup(Context, String, String, String, String, SSLSocketFactory,
+     * cz.msebera.android.httpclient.conn.ssl.SSLSocketFactory)
      */
     public static void setup(Context context, String clientId, String clientSecret) {
```
Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/Podio.java]: is this method used? If not, away with it!

File: sdk/src/main/java/com/podio/sdk/Podio.java [hunk 8]
```diff
@@ -245,33 +249,35 @@ public static void setup(Context context, String clientId, String clientSecret)
...
-        status.setClient(restClient);
+        application.setClient(volleytRestClient);
+        calendar.setClient(volleytRestClient);
+        client.setClient(volleytRestClient);
+        contact.setClient(volleytRestClient);
+        conversation.setClient(volleytRestClient);
+        file.setClient(androidAsyncHttpRestClient);
+        item.setClient(volleytRestClient);
+        organization.setClient(volleytRestClient);
+        user.setClient(volleytRestClient);
```
Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/Podio.java]: I know that is not part of this PR, but some of these guys (eg: `user.setClient(volleytRestClient)`) are also called in the super method.
Do you know the reason for the duplicate?

File: sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java (hunk 3 above)
Comment (tobiaselindberg) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: The listener on line 80 is results only and then errors are returned with the error listener. This is done same way as it was done in volley request.

File: sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java (hunk 2 above)
Comment (tobiaselindberg) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: * We are listening to the refresh and on line 114 when we get onSessionChanged we perform the request then instead.
* I agree on the perform request remark. Will change that.

File: sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java (hunk 6 above)
Comment (tobiaselindberg) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: The difference is that if it's an ApiError then we have a json error from our api that we parse, otherwise we have NoResponseError and ConnectionError and for unknown exceptions we bake them in to the PodioError.

File: sdk/src/main/java/com/podio/sdk/provider/FileProvider.java (hunk 1 above)
Comment (tobiaselindberg) [inline sdk/src/main/java/com/podio/sdk/provider/FileProvider.java]: I looked in our API code and "file/v2" and "file" both point at the same method and I think it looks nicer with just "file" than "file/v2"

File: sdk/src/main/java/com/podio/sdk/Podio.java [hunk 9]
```diff
@@ -245,33 +249,37 @@ public static void setup(Context context, String clientId, String clientSecret)
...
+     *         Optional user agent.
+     * @param volleySslSocketFactory
      *         Optional custom SSL socket factory to use in the HTTP requests.
+     * @param androidAsyncHttpSslSocketFactory
+     *         Optional custom SSL sockey factory to use for uploading files.
      */
-    public static void setup(Context context, String scheme, String authority, String clientId, String clientSecret, SSLSocketFactory sslSocketFactory) {
-        restClient.setup(context, scheme, authority, clientId, clientSecret, sslSocketFactory);
-
+    public static void setup(Context context, String scheme, String authority, String clientId, String clientSecret, String userAgent, SSLSocketFactory volleySslSocketFactory, cz.msebera.android.httpclient.conn.ssl.SSLSocketFactory androidAsyncHttpSslSocketFactory) {
```
Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/Podio.java]: i am confused about this. You changed this signature but I don't see the change reflected in `SuperPodio` caller....

File: sdk/src/main/java/com/podio/sdk/Podio.java (hunk 8 above)
Comment (tobiaselindberg) [inline sdk/src/main/java/com/podio/sdk/Podio.java]: You mean SuperPodio oustide this code base? The Podio android app has a few more privileges than third party developers so the SuperPodio part otuside the SDK handles our special privilges.

File: sdk/src/main/java/com/podio/sdk/Podio.java (hunk 9 above)
Comment (tobiaselindberg) [inline sdk/src/main/java/com/podio/sdk/Podio.java]: What do you mean? It is reflected: https://github.com/podio/android-app/blob/fileupload/app/src/main/java/com/podio/sdk/SuperPodio.java#L128

File: sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java (hunk 4 above)
Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: Definitely is not a requirement from Citrix, we are not doing in the API.
Even in that case, the right way is to do it only once in the Readme, and not for every class.

File: sdk/src/main/java/com/podio/sdk/Podio.java (hunk 8 above)
Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/Podio.java]: Thanks!

File: sdk/src/main/java/com/podio/sdk/Podio.java (hunk 9 above)
Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/Podio.java]: Yup, missed it, my bad

File: sdk/src/main/java/com/podio/sdk/provider/FileProvider.java (hunk 1 above)
Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/provider/FileProvider.java]: :+1:

File: sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java (hunk 4 above)
Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: The same goes for @author comments. That was needed before version control systems.
Now I can see who wrote any line of code and navigate the history. @author is just another annoying comment that I *have to* read before saying "ah, I dont care".

Comment (tobiaselindberg) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: Will you check with Florian about the copy right in the sdk?

Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: yes, when he is back. 
thans!

File: sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java (hunk 2 above)
Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: I guess you forgot, I would like very much to move the performRequest out of the constructor...

File: sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    }
+
+    protected AndroidAsyncHttpRequest(AsyncHttpClient client, Context context, String url, File file, Class<T> resultType) {
+        this.client = client;
+        this.context = context;
+        this.url = url;
+        this.file = file;
+        this.classOfResult = resultType;
+        this.callbackManager = new CallbackManager();
+        this.attemptedReauth = false;
```
Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: what is the purpose of this variable? can only refresh token once?

Comment (tobiaselindberg) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: The purpose is that if we get a 401 because the user e.g. changed their password, then we try once to reauthenticate (See line 154) and if we were not able to refresh then  we know that the refresh token is not doing the job either and the boolean is just to make sure that we don't end up doing an endless loop of 401 -> refreshing -> 401 -> refreshing, etc. A Request is never reused so it wont block a new request from trying to refresh the token.

Comment (jalvz) [inline sdk/src/main/java/com/podio/sdk/androidasynchttp/AndroidAsyncHttpRequest.java]: got it
